### ISSUE TITLE
Signal safety

### ISSUE THREAD

**Issue by skmp** [FEX-Emu/FEX#1682]
Splitting from #1558 & #1677, as well as discussions with @neobrain  and @Sonicadvance1.

## The issues
(a) Signals can interrupt the JIT compiler or syscall or other FEX-related code, which are not guaranteed to be signal re-entrant safe. Any code that touches non-stack memory, or uses mutexes is possibly not signal safe. We currently block signals around some code, either using `ScopedSignalMaskWith*` guards or manually (eg, the dispatcher disabling signal handling around calls to CompileCode)

(b) Signals can interrupt the translated code in the middle of operations that would normally be atomic wrt signals. This may or may not be a problem, depending on how we have implemented x86. A good example is REP* operations. This can be an issue even without LSE elimination, as the recovered guest state might be "teared".

(c) Similar to above, signals can interrupt the translated code in places where we can't recover the guest architectural place, due to optimisations.

## Current status
Our current "signal safety strategy" for (a) is to sprinkle signal disabling code around regions that deadlock. This is very inconsistent throughput the codebase, and there are several bugs waiting to be hit. In general, this is a compromise between "likely to lockup" and "performant code". 

For (b) and (c) we currently only partially recover the guest architectural state, store it alongside the host architectural state, and hope the guest code doesn't care too much about the contents of the guest state, and that it will not modify it. We depend on returning to the interrupted host code using the stored host architectural state, in order to resume execution in the middle of any teared instructions, and eventually exit from some point with a valid guest state. This poses another limitation, that the interrupted block cannot be discarded from the code cache, so the code cache cannot be cleared. This might also have further implications around SMC and code invalidations.

## Proposed solutions
For (a) I'd like us to have clear guidelines on how to handle this, as well as a mode that might be slower but offers guaranteed stability. This needs some thought, but is not too hard.

For (b) and (c) the only viable solution I can think of is a combination of deferring the signal delivery until we have a fully recoverable guest state, and storing metadata that can help us exit from the middle of a block. (c) Can be avoided by limiting store elimination from LSE and disabling DSE. We can have a tradeoff between "defer delay" vs "runtime performance".

Related: #518, #650, #1228, #1666

**Comment by skmp** [comment]
(Updated with more details about signal latency, execution overshoot, and signal queue merging)

**Comment by skmp** [comment]
Apart from blocking signals during critical sections with signal deferring we can handle the signal after the critical section, or at specific signal handling spots. This greatly reduces the complexity and overhead of avoiding deadlocks.

**Comment by skmp** [comment]
Another complication are signals that cannot be blocked or caught, namely SIGKILL and SIGSTOP. While these are not an issue when targeted to the process itself, they may be an issue when/if they can be directed to threads. pthread_kill doesn't allow that to happen, however tgkill/tkill might. Also `pthread_suspend_np & friends`[1] have to be implemented somehow if they exist on linux.

This kind of ties in with our thread / process lifetime and `exit(2)`, `exit_group(2)`, on which I'm not 100% solid on how it works. From a high level perspective, I know we use SIG63 to stop and terminate threads.

### [1] pthread_suspend_np & friends
```
     int pthread_suspend_np(pthread_t thread);
     void pthread_suspend_all_np(void);
     int pthread_resume_np(pthread_t thread);
     void pthread_resume_all_np(void);
```

**Comment by skmp** [comment]
Another edge case that we have to consider is a thunked library registering signals in the host side.

**Comment by skmp** [comment]
(Deferred signals investigation is in https://github.com/FEX-Emu/FEX/issues/1666, will update here with a summary here once that is closed)

**Comment by skmp** [comment]
Another edge case is how to handle cpu state reconstruction around thunks.

Of course, reconstructing context is impossible if we deliver the signal while the thunk is running. Also, depending in the library that is thunked, it might unreasonable for the guest to assume things about the cpu state or the code that is being run. Exceptions are applications that, eg, might try to interpret the opcodes, eg around segfaults, in the signal handler. Another case could be to detect interrupted system calls.

Thunks may perform syscalls, block or execute for an arbitrary amount of time, so deferring delivery to the return to guest could lead to bugs. And synchronous signals cannot be blocked.

A "plausible" state to present to the guest would be, {invalid register context, RIP pointing to the guest-thunk function thunkOp}.

Handling signal delivery in thunks does complicate things a bit for fully deferred signals as it forces us to store the host state along with the guest context.

**Comment by skmp** [comment]
Defining some terms here to make things more understandable.

### `(fully/plausibly/not) redispatchable [guest state]`
A state we can redispatch / recompile and the guest would not be able to detect any changes.

If `fully`, any code can be run.

if `plausibly` the same code that was interrupted during state reconstruction can be resumed, but not any code.

if `not` host state is needed in addition to the guest state to resume execution.

#### `(full/plausible/partial) guest state reconstruction`
A guest state reconstructed from the partial guest state stored in `CPUState`, the host registers as well as any metadata.

`full`  means the guest state is `fully redispatchable`

`plausible` means the guest state is `plausibly redispatchable`

`partial`  means the state is `not redispatchable`.


#### `host deferred signals`
(I've also mentioned this as partial in the past)

Signal delivery is deferred only for host code, except thunks. This avoids reentrancy amplification from our logic, and removes the overhead of using signal masks around locks. It also guarantees that any 3rd party libraries we use directly don't need to have their locks modified - again, thunks are not included here.

#### `guest deferred signals`
Signal delivery is deferred for guest code, so we can do guest state reconstruction.

**Comment by skmp** [comment]
Redzone is applied both for 32-bit and 64-bit processes in a 64-bit kernel. Source: https://elixir.bootlin.com/linux/latest/source/arch/x86/kernel/signal.c#L252

**Comment by skmp** [comment]
https://elixir.bootlin.com/linux/latest/source/arch/x86/um/signal.c#L481

the `rt_sigframe` has special some meaning for some applications (eg, gdb) according to comments around the kernel, we could match ours with it for guest gdb debugging.

**Comment by skmp** [comment]
One more note on the actual flow in the linux kernel.

For non usermode-linux kernels, signal delivery starts from `handle_signal_work` (https://elixir.bootlin.com/linux/latest/source/kernel/entry/common.c#L143) which calls `arch_do_signal_or_restart` (https://elixir.bootlin.com/linux/latest/source/include/linux/entry-common.h#L280) implemented in https://elixir.bootlin.com/linux/latest/source/arch/x86/kernel/signal.c#L864.

`arch_do_signal_or_restart` calls `handle_signal` (https://elixir.bootlin.com/linux/latest/source/arch/x86/kernel/signal.c#L786) and returns if there is a signal to handle, or handles automatic syscall restart and calls `restore_saved_sigmask` and exits.

`handle_signal` (https://elixir.bootlin.com/linux/latest/source/arch/x86/kernel/signal.c#L786) first handles v8086 mode (we don't care), then handles syscall restart, then handles single stepping, then setups the frame with `setup_rt_frame`, and if there are no errors there, it clears DF, RF, TR and the fpu state. Finally, `signal_setup_done` is called.

`setup_rt_frame` (https://elixir.bootlin.com/linux/latest/source/arch/x86/kernel/signal.c#L763) `ia32_setup_rt_frame` (compat task, `SA_SIGINFO`) or ia32_setup_frame (compat task, no `SA_SIGINFO`) or `__setup_rt_frame` (x86-64 task)

`signal_setup_done` (https://elixir.bootlin.com/linux/latest/source/kernel/signal.c#L2905) calls `force_sigsegv` (https://elixir.bootlin.com/linux/latest/source/kernel/signal.c#L1697) on error or `signal_delivered` (https://elixir.bootlin.com/linux/latest/source/kernel/signal.c#L2886)

`force_sigsegv` (https://elixir.bootlin.com/linux/latest/source/kernel/signal.c#L1697)  sends a non-fatal sigsegv if not in a sigsegv loop. I haven't fully followed that path, and will do so, but I expect no surprises.

`signal_delivered`(https://elixir.bootlin.com/linux/latest/source/kernel/signal.c#L2886) handles `SA_NODEFER`, `SS_AUTODISARM`

#### syscall restarting
The syscall number is saved in `regs->orig_ax` (`syscall_get_nr`, https://elixir.bootlin.com/linux/latest/source/arch/x86/include/asm/syscall.h#L38). Interestingly enough, `syscall_rollback` is not used during syscall restart, but manually reimplemented `regs->ax = regs->orig_ax;`.





From there on, `setup_signal_stack_si` (https://elixir.bootlin.com/linux/latest/source/arch/x86/um/signal.c#L361)

**Comment by skmp** [comment]
#### Complication: Signal mask handling vs thunks

A host thunk may modify the signal mask, and the guest will not be informed about, so if it goes `getprocmask` it will get an unexpected value.

We could synchronize the masks upon entering guest after a thunk (either recursed, or returning), and making sure we forward the masking to the thunk.

This may not be 100% correct as we need to always have a couple of signals unblocked.

**Comment by skmp** [comment]
#### Complication: Signal mask handling on guest signal returns

When returning from a guest signal handler we need to give the guest's signal mask back to the kernel, not the host's. A guest application can modify the signal mask stored in `ucontext`.

**Comment by skmp** [comment]
#### Note: X86ContextBackup/ArmContextBackup may not contain or restore all of the context

**Comment by skmp** [comment]
Run into guest stack overflows in a sample app that used 4kb stacks, both with `main` and `skmp/guest-rt_sigreturn`.

**Comment by skmp** [comment]
Another interesting tidbit
```
POSIX 3.3.1.3:
 "Setting a signal action to SIG_IGN for a signal that is
  pending shall cause the pending signal to be discarded,
  whether or not it is blocked."
 "Setting a signal action to SIG_DFL for a signal that is
  pending and whose default action is to ignore the signal
  (for example, SIGCHLD), shall cause the pending signal to
  be discarded, whether or not it is blocked"
```

**Comment by Sonicadvance1** [comment]
Watch out with how signal queueing has changed how this has worked a bit.

**Comment by skmp** [comment]
Yes, I'm not trusting anything (except how the kernel actually implements things) at this point

**Comment by skmp** [comment]
And everything is subject to a test case

**Comment by skmp** [comment]
Note: As a general design goal, it would be nice to more more (most?) of the guest signal handling to `syscalls` from the `jit dispatcher`

**Comment by skmp** [comment]
sigprocmask is not working in the current implementation at all. It blocks other signals during signal delivery to host, not during signal delivery to guest.

**Comment by skmp** [comment]
`GuestSigAction`: GuestSAMask doesn't match kernel structure padding. Kernels are configured by default with only 64 signals, so this should be ok, but they can be configured with up to 1024 signals.

**Comment by skmp** [comment]
Note from https://man7.org/linux/man-pages/man2/sigaction.2.html

```
   Undocumented
       Before the introduction of SA_SIGINFO, it was also possible to
       get some additional information about the signal.  This was done
       by providing an sa_handler signal handler with a second argument
       of type struct sigcontext, which is the same structure as the one
       that is passed in the uc_mcontext field of the ucontext structure
       that is passed (via a pointer) in the third argument of the
       sa_sigaction handler.  See the relevant Linux kernel sources for
       details.  This use is obsolete now.
```

**Comment by skmp** [comment]
Note: There are several flags of `sigaction` that we may not support correctly right now (`SA_EXPOSE_TAGBITS`, `SA_RESETHAND`, `SA_RESTORER`, `SA_NODEFER`)

**Comment by skmp** [comment]
Note: pselect6 (and possibly others) take the signal mask as a parameter, which might have complications around host -> guest signal handover.

**Comment by skmp** [comment]
Issue: Our internal signals that shouldn't be blocked (SIGRT31, any others?) will pass a guest-sent, guest-blocked signal to the guest, instead of holding it.

We need to at least defer in our signal handling and not deliver to the guest, though can can get tricky, and is probably leaky wrt sigwait and other system calls.

At least `conformance-interfaces-sigwaitinfo-2-1.test.jit.posix` triggers this, though the test itself ignores the issue.

From what I can see, there's no simple way (and maybe none at all) for us to "safely" steal or overload signals from the guest.

I did a quick search for alternative solutions, and possibly ptracing outselves is a cleaner solution to all this, though that needs a deeper investigation.

**Comment by skmp** [comment]
Issue: glibc uses internally signals 32, 33 (and possibly 34, only? in LinuxThreads which is no longer used). Guest glibc and host glibc handling might cause issues there. We need test cases.

**Comment by skmp** [comment]
Note: Host-thunks probably should use the guest signal interface and delivery, to present a consistent view to the guest.

**Comment by skmp** [comment]
```The [POSIX realtime signals](http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_04_02) option defines a set of signals from SIGRTMIN to SIGRTMAX which have various useful properties (e.g. they have a well-defined delivery priority -- lowest signal number first -- and multiple instances of the same signal can be queued, and associated with a parameter, via sigqueue()). These are implemented by the kernel using signal numbers 32 upwards.```

**Comment by skmp** [comment]
Note: `After returning from a signal-catching function, the value of errno is unspecified if the signal-catching function or any function it called assigned a value to errno and the signal-catching function did not save and restore the original value of errno. `

**Comment by skmp** [comment]
For some timer details,

https://man7.org/linux/man-pages/man2/timer_create.2.html

Likely to be those two
```

**Comment by skmp** [comment]
Digging deeper in glibc,

It seems to create a helper thread (
https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/sysdeps/unix/sysv/linux/timer_routines.c#L62)

which does
```while (__sigwaitinfo (&sigtimer_set, &si) < 0);```
and then
```if (si.si_code == SI_TIMER)```
and then creates a new thread to run the callback if it is found.

The only other users are `__libc_signal_block_sigtimer` (never used) and `__libc_signal_unblock_sigtimer` (used in the second helper thread, https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/sysdeps/unix/sysv/linux/timer_routines.c#L44, no idea why, *looks like a bug* / not needed?)

The thread itself is created in `___timer_create` (https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/sysdeps/unix/sysv/linux/timer_create.c#L70)

It is re-set to be re-created post-fork (https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/sysdeps/unix/sysv/linux/timer_routines.c#L118)

called via `fork_system_setup_after_fork` https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/sysdeps/nptl/fork.h#L48

as only the forking thread makes it across the fork.

**Comment by skmp** [comment]
from https://man7.org/linux/man-pages/man2/sigwaitinfo.2.html

````In a multithreaded program, the signal should be
       blocked in all threads, in order to prevent the signal being
       treated according to its default disposition in a thread other
       than the one calling sigwaitinfo() or sigtimedwait()).```

That `__libc_signal_unblock_sigtimer` looks like a glibc bug.

**Comment by skmp** [comment]
We correctly forward SIG32 to the guest (wrote two tests about it), and it should be safe to use posix timers with `SIGEV_THREAD` from within FEX or any host thunks, as glibc targets by tid for it.

We cannot use safely `pthread_cancel` towards a thread that is also a guest thread, as it will cause a guest abort, not a host abort, and any host thread-level destructors will not be called.

While not tested yet, SIG33 should work fine for the guest, however doing `setuid` from FEX or a thunk is likely to not work correctly. 

SIG33 is sent  through `INLINE_SETXID_SYSCALL` https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/sysdeps/nptl/setxid.h#L29 (only on multi threaded programs)

via `__nptl_setxid` (https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/nptl/nptl_setxid.c#L175)

and finally `setxid_signal_thread`
(https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/nptl/nptl_setxid.c#L154)

and received with `__nptl_setxid_sighandler`
(https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/nptl/nptl_setxid.c#L56)


While a bit ugly, we could piggyback on SIG33, as it is (almost) never blocked, and use `SI_TIMER`, as only `SI_TKILL` is handled by glibc.

**Comment by skmp** [comment]
Looking in detail in our implementation, 

#### `SignalEvent::Stop` is used to
- Implement the `exit` syscall (via `Context::StopThread`)
- Implement OP_Break in the interpreter  (via `Context::StopThread`)
- Implement graceful shutdown for `exit_group` (via `Context::Stop`, has a race bug)
- Implement graceful shutdown for Debugger (via `Context::Stop` in `CloseCallback` and `ExitHandler`)

#### `SignalEvent::Pause` is used 
- To implement `PauseCallback` in `Debugger/Main.cpp` (via `Context::Pause`)
- In several places in `GdbServer.cpp` to pause all threads (via `Context::Pause`)

#### `SignalEvent::Return` is used
- in `Context::Run` (? why?) Which is used by `GdbServer.cpp`
- In `Context::Step` via `Context::Run`, which also uses `CoreRunningMode::MODE_SINGLESTEP` which is broken, as the blocks check on entry to early exit, not on exit.
- In `SignalReturn` of `Interpreter/BranchOps.cpp` to implement signal returns (this seems the only intended use)

JIT signal returns use either `ud2` or `hlt(0)` in the dispatcher

### Can we avoid using signals there?

- All uses of `SignalEvent::Stop` can be switched over to use the dispatcher
- `SignalEvent::Pause` is more tricky, though debugger only. We could pause threads via a change of `CoreRunningMode`, which is checked on every block when gdb is attached, though this would still block on syscalls. In a more complex solution, the debugger could ptrace FEX, which has other advantages as well.
- `SignalEvent::Return` can be switched over to use the dispatcher, and/or even become not needed by implementing `rt_sigreturn` and `sigreturn`

I'll look into removing all uses, apart from `SignalEvent::Pause` for debugger.

**Comment by skmp** [comment]
We need to handle SIGSEGV, SIGBUS internally, which means if the guest masks them, we handle them incorrectly. 

SIGSEGV, SIGBUS, SIGILL, SIGFPE all terminate the process based on my testing if masked and generated synchronously. They obey normal queueing rules otherwise (kill, t/tgkill, sigqueue, timer?)

**Comment by skmp** [comment]
Interesting read: http://davmac.org/davpage/linux/rtsignals.html

```
Well, with kernel 2.6.22.6 at least, you *can* have two of the same signal pending: one queued via sigqueue (to the process) plus one via raise or pthread_kill (to the thread). The thread-specific signal gets delivered first).
```

**Comment by skmp** [comment]
`/proc/{pid}/status` and ``/proc/self/status` include
```
SigQ:	1/192451
SigPnd:	0000000000000000
ShdPnd:	0000000000000000
SigBlk:	7be3c0fe28014a03
SigIgn:	0000000000001000
SigCgt:	00000001000004ec
```

Something to be mindful of
